src/digests.rs

```diff
@@ -0,0 +1,211 @@
+        if input.is_empty() {
+            return Err(DecodeError::BadInputLength);
+        }
+
+        let (_code, bytes) = varint_decode::u64(&input).map_err(|_| DecodeError::BadInputLength)?;
+
+        let (hash_len, bytes) =
+            varint_decode::u64(&bytes).map_err(|_| DecodeError::BadInputLength)?;
+        if (bytes.len() as u64) != hash_len {
+            return Err(DecodeError::BadInputLength);
+        }
+
+        Ok(MultihashRef { bytes: input })
+    }
+
+    /// Returns which hashing algorithm is used in this multihash.
+    pub fn algorithm(&self) -> Code {
+        let code = varint_decode::u64(&self.bytes)
+            .expect("multihash is known to be valid algorithm")
+            .0;
+        Code::from_u64(code).expect("multihash is known to be valid")
+    }
+
+    /// Returns the hashed data.
+    pub fn digest(&self) -> &'a [u8] {
+        let bytes = varint_decode::u16(&self.bytes)
+            .expect("multihash is known to be valid digest")
+            .1;
+        &bytes[1..]
+    }
+
+    /// Builds a `Multihash` that owns the data.
+    ///
+    /// This operation allocates.
+    pub fn to_owned(&self) -> Multihash {
+        Multihash {
+            bytes: Bytes::copy_from_slice(self.bytes),
+        }
+    }
+
+    /// Returns the bytes representation of this multihash.
+    pub fn as_bytes(&self) -> &'a [u8] {
+        &self.bytes
+    }
+}
+
+impl<'a> PartialEq<Multihash> for MultihashRef<'a> {
+    fn eq(&self, other: &Multihash) -> bool {
+        self.bytes == &*other.bytes
+    }
+}
+
+/// The `MultihashDigest` trait specifies an interface common for all multihash functions.
+pub trait MultihashDigest {
+    /// The Mutlihash byte value.
+    const CODE: u64;
+
+    /// Hash some input and return the digest.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the digest length is bigger than 2^32. This only happens for identity hasing.
+    fn digest(data: &[u8]) -> Multihash;
+
+    //fn dyn_digest(&self, data: &[u8]) -> Multihash {
+    //    Self::digest(data)
+    //}
+}
+
+/// The `DynMultihashDigest` trait is a variant of the `MultihashDigest` that can be used as trait
+/// object.
+pub trait DynMultihashDigest {
+    /// The Mutlihash byte value.
+    fn code(&self) -> u64;
+
+    /// Hash some input and return the digest.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the digest length is bigger than 2^32. This only happens for identity hasing.
+    fn digest(&self, data: &[u8]) -> Multihash;
+}
+
+impl<T: MultihashDigest + ?Sized> DynMultihashDigest for T {
+    fn code(&self) -> u64 {
+        Self::CODE
+    }
+    fn digest(&self, data: &[u8]) -> Multihash {
+        Self::digest(data)
+    }
+}
+
+/// Wraps a hash digest in Multihash with the given Mutlihash code.
+///
+/// The size of the hash is determoned by the size of the input hash. If it should be truncated
+/// the input data must already be the truncated hash.
+pub fn wrap(code: u64, data: &[u8]) -> Multihash {
```

Comment:
This should probably take a `Code`, not a `u64`

Reply:
The reason why this doesn't take a `Code` is, that it can be used for your custom, not yet implemented Multihash, without the need to fork the library. I thought about making it a trait, but that seemed a bit like overkill.

Reply:
@vmx I would suggest having a separate `fn` for that use case and have primary one be `Code` as that makes it more clear.

Reply:
Or better yet might be adding `Custom(u64)` to the enum so that use of non standard codes is explicit.